sysutils/lambd/files/lambd.8
.\"	$NetBSD: lambd.8,v 1.3 2001/07/18 06:47:37 itojun Exp $
.\"
.\" Copyright (C) 2001 WIDE Project.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the project nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE PROJECT AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE PROJECT OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd July 14, 2001
.Dt LAMBD 8
.Os
.Sh NAME
.Nm lambd
.Nd shutdown switch monitoring daemon for Wildlab LAMB
.Sh SYNOPSIS
.Nm ""
.Op Fl dfn
.Op Ar usec
.Sh DESCRIPTION
.Nm
monitors shutdown switch on Wildlab LAMB, and issues
.Dq Li shutdown -h now
if pressed.
It also blinks LED every 1 second, or the interval specified by
.Ar usec .
Normally
.Nm
becomes a daemon on invocation.
.Pp
Three command line options,
.Fl dfn
are supplied for debugging purposes.
.Fl d
lets it generate more messages, and forbids it from shutting down the machine.
.Fl f
runs
.Nm
in foreground mode.
.Fl n
avoids any hardware I/O accesses from
.Nm Ns .
.Sh SEE ALSO
.Pa http://www.wildlab.com/
.Sh BUGS
.Nm
should use LED for more interesting purposes,
like indicating load average, morse code and such.


doc/NEWS
# $NetBSD: NEWS,v 1.33 2025/12/03 21:57:26 leot Exp $

PKGSRC NEWS
===========

  This file contains information about important infrastructural changes in
  pkgsrc. The intended readers are the pkgsrc developers as well as anyone
  tracking pkgsrc-current.

m68k softfloat (m68ksf) support [nat 2025-10-22]

        Support for m68k softfloat was added to pkgsrc.

Default versions bumped [wiz 2025-10-06]

	PYTHON_VERSION_DEFAULT is now 313.

Default MySQL implementation changed to MariaDB 11.4 [nia 2024-12-25]

	The default version of MySQL as set by the pkgsrc infrastructure
	was changed from MariaDB 10.11 to MariaDB 11.4.

Default versions bumped [wiz 2024-08-12]

	PYTHON_VERSION_DEFAULT is now 312.

Default versions bumped [taca 2024-02-10]

	RUBY_VERSION_DEFAULT is now 32.

Default MySQL implementation changed to MariaDB 10.11 [nia 2024-01-23]

	The default version of MySQL as set by the pkgsrc infrastructure
	was changed from MariaDB 10.6 to MariaDB 10.11.

Default versions bumped [wiz 2023-08-14]

	PYTHON_VERSION_DEFAULT is now 311.

Support for USE_CC_FEATURES and USE_CXX_FEAUTRES [nia 2023-07-18]

	USE_CC_FEATURES and USE_CXX_FEATURES can now be used
	to specify a minimum level of C and C++ support required
	to compile a package, including specific libstdc++ features.

	ALLOW_NEWER_COMPILER can be set to no in mk.conf to skip
	building packages that require a newer compiler than the system
	provides.

Default versions bumped [nia 2022-06-30]

	PYTHON_VERSION_DEFAULT is now 310.
	PGSQL_VERSION_DEFAULT is now 14.
	LUA_VERSION_DEFAULT is now 54.

Dropped support for compiling glib2-using packages without introspection
support [wiz 2022-06-07]

PKGSRC_USE_MKTOOLS support added [jperkin 2022-06-07]

        Added C-based tools to speed up various parts of the
        mk infrastructure via PKGSRC_USE_MKTOOLS=yes in mk.conf.

        For packages with large numbers of dependencies, you should see
        anywhere up to around 10x speed-up for the 'wrapper' phase, and
        perhaps more importantly a huge reduction in system time usage.

Linux infrastructure adapted to non-standard distributions [nia 2022-05-01]

        The Linux pkgsrc infrastructure was adapted to support Linux
        distributions where binaries and shared objects are not
        stored in a normal Unix-like file system layout.  It's hoped
        that this will be useful for pkgsrc development on platforms
        like NixOS and Gobo Linux.

UnixWare support [nia 2021-12-26]

        Support for SCO/Xinuos UnixWare was improved to the point where
        it is possible to complete a bootstrap on this platform.

SHA1 distfile hashes deprecated [nia 2021-10-07]

        Distfile hash algorithms have been changed, keeping SHA512,
        adding BLAKE2s, and removing SHA1 and RMD160.  New packages
        and updates will only generate SHA512 and BLAKE2s, and the
        stored hashes have been adjusted.

Default Python version changed to 3.9 [nia 2021-10-01]

Increased enabled-by-default hardening options [nia 2021-09-27]

	PKGSRC_USE_MKPIE was enabled on some architectures.
	Packages that provide static libraries to other packages
	should be rebuilt with MKPIE enabled.
	Some packages may experience runtime errors on platforms
	like i386 - usually this is due to unsafe assembly code,
	which should be disabled when MKPIE is enabled.

	PKGSRC_USE_SSP was bumped from "yes" to "strong".  This
	is not expected to have a noticable effect.

	More information:
	https://www.NetBSD.org/docs/pkgsrc/hardening.html
	(or see "Security hardening" section of doc/pkgsrc.txt)

Default MySQL implementation changed to MariaDB 10.6 [nia 2021-09-27]

Default PostgreSQL changed to 13 [nia 2021-06-28]

Default MySQL implementation changed to MariaDB [nia 2021-06-24]

	The default version of MySQL as set by the pkgsrc infrastructure
	was changed from MySQL 5.7 to MariaDB 10.5.

	This is configurable by setting MYSQL_VERSION_DEFAULT in mk.conf.

	PHP packages are no longer tied to a specific MySQL version
	due to the native driver being enabled by default.

Python 3.8 default [nia 2020-12-04]
	The default Python version was switched to 3.8.

Default database location changed for NetBSD [wiz 2020-12-02]
	The database of installed packages now defaults to ${PREFIX}/pkgdb
	instead of /var/db/pkg on NetBSD. This matches the default on
	other operating systems and cases where a non-default bootstrap
	is used.

	To deal with this change, see this blog post:

	https://pkgsrc.org/news/pkgdb-change/

	Since this change affects the packaging tools only, there is no change
	to binary packages and they do not need to be rebuilt.

lua 5.3 default [nia 2020-03-30]
	The default Lua version was switched to 5.3.
	Since many packages specify a preferred Lua version, this is mainly
	expected to affect users of Lua modules who were using the default
	settings.

gfortran default [nia 2020-03-30]
	gfortran is now the default Fortran compiler instead of g95.
	gfortran is expected to support more pkgsrc platforms and be
	better at compiling newer Fortran code.

python 3.7 default [maya 2019-04-24]
	As Python 2.x versions are nearing end of life, we have switched
	the default to a newer version.
	Python 2.7 packages are still available, and most python packages
	are built for all available python versions.

Lua multiversion support [alnsn 2014-05-03]
	Lua packages lang/lua51 and lang/lua52 have been modified
	to coexist without conflicts, infrastructure for supporting
	multiple versions of Lua has been put in place and all
	dependent packages have been adapted to the new infrastructure.

SMF support is introduced [jperkin 2014-03-11]

Unprivileged staged installation is default [asau 2012-08-14]
	pkgsrc has changed to use unprivileged staged installation by default.
	This makes it possible to prevent some installation mistakes.
	It also makes detection of package conflicts easier.
	Previously you had to set PKG_DESTDIR_SUPPORT to "user-destdir"

sysutils/lambd/files/lambd.c
/*	$NetBSD: lambd.c,v 1.3 2001/07/19 20:23:43 itojun Exp $	*/

/*
 * Copyright (C) 2001 WIDE Project.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the project nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE PROJECT AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE PROJECT OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/types.h>
#include <machine/sysarch.h>
#include <machine/pio.h>
#include <stdlib.h>
#include <unistd.h>
#include <stdio.h>
#include <err.h>
#include <ctype.h>

#include "lambio.h"

#define BASEPORT	(0x378)

int main __P((int, char **));
void usage __P((void));
void mainloop __P((void));
void lsleep __P((unsigned long));
int monitor __P((void));
void led __P((int));

int cycle __P((void));
int loadavg __P((void));
int morse __P((void));

int foreground = 0;
int debug = 0;
#ifdef __i386__
int nohw = 0;
#else
const int nohw = 1;
#endif
unsigned long delay = 1 * 1000000;
const char *morsestr = "LAMB";

int (*func) __P((void)) = cycle;

int
main(argc, argv)
	int argc;
	char **argv;
{
	int ch;
	char *p, *ep;

	while ((ch = getopt(argc, argv, "LM:dfn")) != -1) {
		switch (ch) {
		case 'L':
			func = loadavg;
			break;
		case 'M':
			func = morse;
			morsestr = optarg;
			break;
		case 'd':
			debug++;
			break;
		case 'f':
			foreground++;
			break;
		case 'n':
			nohw++;
			break;
		default:
			usage();
			exit(1);
		}
	}

	if (!islamb()) {
		if (debug)
			fprintf(stderr, "it is not lamb\n");
		nohw++;
	} else
		if (debug)
			fprintf(stderr, "it is indeed lamb\n");

	argc -= optind;
	argv += optind;

	switch (argc) {
	case 0:
		break;
	case 1:
		p = *argv;
		ep = NULL;
		delay = strtoul(*argv, &ep, 10);
		if (ep && !*ep)
			break;
		/* FALLTHROUGH */
	default:
		usage();
		exit(1);
	}

	if (!nohw) {
		if (lamb_open() < 0) {
			err(1, "lamb_open");
			/* NOTREACHED */
		}
	}

	if (!foreground)
		daemon(0, 0);

	if (debug)
		fprintf(stderr, "delay=%lu\n", delay);
	mainloop();
	/* NOTREACHED */

	exit(0);	/* silent gcc */
}

void
usage()
{
	fprintf(stderr, "usage: lambd [-LMdfn] [usec]\n");
}

void
mainloop()
{
	led(0);
	while (1) {
		if ((*func)())
			break;
	}

	led(0);
	if (debug) {
		fprintf(stderr, "shutdown -h now\n");
		exit(0);
	} else {
		execl("/sbin/shutdown", "shutdown", "-h", "now", NULL);
		/* NOTREACHED */
	}
}

void
lsleep(usec)
	unsigned long usec;
{

	sleep(usec / 1000000);
	usleep(usec % 1000000);
}

int
monitor()
{
	int i;

	if (nohw)
		return 0;

	for (i = 0; i < 10; i++)
		if (!lamb_reboot())
			return 0;
	return 1;
}

void
led(on)
	int on;
{

	if (debug)
		fprintf(stderr, "led=%d\n", on);
	if (nohw)
		return;

	lamb_led(on);
}

/*
 * define a interesting functions below!
 */

int
cycle()
{
	if (monitor())
		return 1;
	led(1);
	lsleep(delay / 2);

	if (monitor())
		return 1;
	led(0);
	lsleep(delay / 2);
	return 0;
}

int
loadavg()
{
	double v;
	int i;
	unsigned long t = delay / 10;

	getloadavg(&v, 1);
	if (debug)
		fprintf(stderr, "loadavg=%g\n", v);
	for (i = 0; i < v * 2; i++) {
		if (monitor())
			return 1;
		led(1);
		lsleep(t);
		led(0);
		lsleep(t);
	}
	lsleep(t * 5);
	return 0;
}

static const char *const digit[] = {
	"-----",
	".----",
	"..---",
	"...--",
	"....-",
	".....",
	"-....",
	"--...",
	"---..",
	"----.",
};
static const char *const alph[] = {
	".-",
	"-...",
	"-.-.",
	"-..",
	".",
	"..-.",
	"--.",
	"....",
	"..",
	".---",
	"-.-",
	".-..",
	"--",
	"-.",
	"---",
	".--.",
	"--.-",
	".-.",
	"...",
	"-",
	"..-",
	"...-",
	".--",
	"-..-",
	"-.--",
	"--..",
};

int
morse()
{
	const char *p, *q;
	int mark, blank;
	unsigned long t = delay / 10;

	for (p = morsestr; *p; p++) {
		if (isdigit(*p))
			q = digit[*p - '0'];
		else if (isalpha(*p) && isupper(*p))
			q = alph[*p - 'A'];
		else if (isalpha(*p) && islower(*p))
			q = alph[*p - 'a'];

		if (debug)
			printf("%c <%s>\n", *p, q);
		for (/*nothing*/; *q; q++) {
			switch (*q) {
			case '-':
				mark = 3; blank = 1; break;
			case '.':
				mark = 1; blank = 1; break;
			default:
				mark = 0; blank = 3; break;
			}
			led(1);
			lsleep(mark * t);
			led(0);
			lsleep(blank * t);
		}
		lsleep(1 * t);
	}
	lsleep(10 * t);
	return 0;
}
